### PR TITLE
Compress metric units defensively

### PR DESCRIPTION
**Pull request by blazember**
The units of the metrics are directly derived from the `ProbeUnit` enum
and the `MetricCompressor` takes the ordinal when creating the blob. On
the consumer side, this ordinal is taken and a lookup is made with the
ordinal potentially leading to `IndexOutOfBoundsException` if the ordinal
is higher than the size of the array holding the possible values of the
units. This can be the case every time we add a new unit to the
`ProbeUnit` enum, which the `MetricCompressor` is not aware of.

To prevent potential issues, we make the unit compression and extraction
more defensive. We change the compression to convert the marked units to
a tag with the name `"metric-unit"` and set the unit to `null`. With this, the
metric can still be processed by MC and the consumers using a metric
extractor and retain the information about the unit. The extractor is
made more defensive by adding a bound check to prevent the
`IndexOutOfBoundsException` in the mentioned case.

### CODE REVIEW

**Comment by blazember** [comment]
@emre-aydin Can you check MC with this change?

**Comment by emre-aydin** [comment]
@blazember doesn't seem to make any difference.
```
java.lang.ArrayIndexOutOfBoundsException: Index 7 out of bounds for length 7
	at com.hazelcast.internal.metrics.impl.MetricsCompressor$MetricsDecompressor.fillUnit(MetricsCompressor.java:546) ~[hazelcast-management-center-4.2020.11-SNAPSHOT.jar:?]
	at com.hazelcast.internal.metrics.impl.MetricsCompressor$MetricsDecompressor.readMetricDescriptor(MetricsCompressor.java:502) ~[hazelcast-management-center-4.2020.11-SNAPSHOT.jar:?]
	at com.hazelcast.internal.metrics.impl.MetricsCompressor$MetricsDecompressor.extractMetrics(MetricsCompressor.java:473) ~[hazelcast-management-center-4.2020.11-SNAPSHOT.jar:?]
	at com.hazelcast.internal.metrics.impl.MetricsCompressor$MetricsDecompressor.access$100(MetricsCompressor.java:450) ~[hazelcast-management-center-4.2020.11-SNAPSHOT.jar:?]
	at com.hazelcast.internal.metrics.impl.MetricsCompressor.extractMetrics(MetricsCompressor.java:383) ~[hazelcast-management-center-4.2020.11-SNAPSHOT.jar:?]
	at com.hazelcast.internal.metrics.impl.MetricsCompressor.extractMetrics(MetricsCompressor.java:356) ~[hazelcast-management-center-4.2020.11-SNAPSHOT.jar:?]
	at com.hazelcast.webmonitor.service.metrics.MetricsConsumer.consumeImdgCollection(MetricsConsumer.java:151) ~[hazelcast-management-center-4.2020.11-SNAPSHOT.jar:?]
	at com.hazelcast.webmonitor.service.metrics.MetricsConsumer.access$500(MetricsConsumer.java:41) ~[hazelcast-management-center-4.2020.11-SNAPSHOT.jar:?]
	at com.hazelcast.webmonitor.service.metrics.MetricsConsumer$ImdgCollectionConsumer.apply(MetricsConsumer.java:266) ~[hazelcast-management-center-4.2020.11-SNAPSHOT.jar:?]
	at com.hazelcast.webmonitor.service.metrics.MetricsConsumer.lambda$readMetrics$2(MetricsConsumer.java:109) ~[hazelcast-management-center-4.2020.11-SNAPSHOT.jar:?]
	at com.hazelcast.client.impl.ClientDelegatingFuture$WhenCompleteAdapter.apply(ClientDelegatingFuture.java:486) ~[hazelcast-management-center-4.2020.11-SNAPSHOT.jar:?]
	at com.hazelcast.client.impl.ClientDelegatingFuture$WhenCompleteAdapter.apply(ClientDelegatingFuture.java:473) ~[hazelcast-management-center-4.2020.11-SNAPSHOT.jar:?]
	at com.hazelcast.client.impl.spi.impl.ClientInvocationFuture$CallIdTrackingBiFunction.apply(ClientInvocationFuture.java:250) ~[hazelcast-management-center-4.2020.11-SNAPSHOT.jar:?]
	at com.hazelcast.spi.impl.AbstractInvocationFuture$HandleNode.lambda$execute$0(AbstractInvocationFuture.java:1490) ~[hazelcast-management-center-4.2020.11-SNAPSHOT.jar:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) [?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:264) [?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
	at java.lang.Thread.run(Thread.java:834) [?:?]
```

**Review by emre-aydin — APPROVED** [review]
Fix looks good to me. I have also tested the change with MC and it works. In MC, we should convert incoming metrics with `metric-unit` tag to unit so that our users will not need to care for two different tags. I will create an issue for this on MC repo.

**Review by puzpuzpuz — APPROVED** [review]

File: hazelcast/src/main/java/com/hazelcast/internal/metrics/ProbeUnit.java [hunk 1]
```diff
@@ -36,4 +34,26 @@
     BOOLEAN,
     /** 0..n, ordinal of an enum */
     ENUM,
+    /** Timestamp or duration represented in µs */
+    // RU_COMPAT_4_2
+    // remove constructor argument in 4.3
```
[the remark this answers is not in the document]
Comment (blazember) [reply hazelcast/src/main/java/com/hazelcast/internal/metrics/ProbeUnit.java:39]: I think if MC handles the units as @emre commented, we can delete the argument safely.

[the remark this answers is not in the document]
Comment (emre-aydin) [reply hazelcast/src/main/java/com/hazelcast/internal/metrics/ProbeUnit.java:39]: I considered the same question and I think yes. We will upgrade MC to latest minor release when it comes out.

[the remark this answers is not in the document]
Comment (puzpuzpuz) [reply hazelcast/src/main/java/com/hazelcast/internal/metrics/ProbeUnit.java:39]: Thanks for the clarification guys!

File: hazelcast/src/main/java/com/hazelcast/internal/metrics/ProbeUnit.java [hunk 2]
```diff
@@ -36,4 +34,26 @@
...
+    US(true);
```
[the remark this answers is not in the document]
Comment (blazember) [reply hazelcast/src/main/java/com/hazelcast/internal/metrics/ProbeUnit.java:57]: Thanks for the comment, added the above roles to the javadoc of the enum.

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/3097/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/3097/Hazelcast-pr-builder-3097.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   ReplicatedMapEntryListenerOnReconnectTest>AbstractListenersOnReconnectTest.testListenersTerminateCluster_smart_multipleServer:246->AbstractListenersOnReconnectTest.testListenersTerminateCluster:289->HazelcastTestSupport.assertClusterSizeEventually:1053->HazelcastTestSupport.assertClusterSizeEventually:1065->HazelcastTestSupport.assertTrueEventually:1270->HazelcastTestSupport.assertTrueEventually:1253->HazelcastTestSupport.assertClusterSize:1040 Cluster size is not correct. Expected: 3, actual: 1, instance index: 0
[INFO] 
[ERROR] Tests run: 35388, Failures: 1, Errors: 0, Skipped: 979
[INFO] 

[ERROR] There are test failures.
</details>
</pre>

**Comment by blazember** [comment]
run-lab-run

**Comment by blazember** [comment]
Thanks for the review @emre-aydin @puzpuzpuz :+1:

**Pull request merged**